### PR TITLE
Make JNIBridge build on Windows

### PR DESCRIPTION
**Pull request by todi1856**
Since I am Windows guy, I didn't want to develop JNIBridge on OSX. So made it buildable on Windows.

**Note:** When producing builds.zip, you would still use OSX build config on Yamato.

Things done:
* You can now build JNIBridge on Windows
* Added Yamato config which validates that JNIBridge builds on Windows
* Added ability for Bee to generate Visual Studio project for JNIBridge and JNIBridgeTests, which makes development much much nicer. Plus C++ debugger works!!!!!
* Yamato Test Trigger now checks all configs that they build correctly
* Expanded build.pl with helpful commands you wouldn't need to memorize how to run stuff
* Replaced size_t to jsize, since I was getting a bunch of 'argument': conversion from 'size_t' to 'jsize', possible loss of data', and since JNIEnv uses jsize, I think it's safe for us to use it as well
* In test code I added a bunch of
```
	if (env->ExceptionOccurred())
	{
		env->ExceptionDescribe();
		exit(-1);
	}

```
otherwise tests on Windows don't fail, and produces false positives during JNI errors

If you run build.pl without anything you'll get this menu:

```
(b) Build JNIBridge and Zip It
(b armeabi-v7a) Build JNIBridge armeabi-v7a
(b arm64-v8a) Build JNIBridge arm64-v8a
(b x86) Build JNIBridge x86
(b x86_64) Build JNIBridge x86_64
(p) Generate Visual Studio project files
(t) Test JNIBridge (only works on OSX)
(h) Show command line arguments
(q) Exit
```

h - command prints this
```
build.pl jnibridge
    Build JNIBridge and Zip It

build.pl jnibridge armeabi-v7a
    Build JNIBridge armeabi-v7a

build.pl jnibridge arm64-v8a
    Build JNIBridge arm64-v8a

build.pl jnibridge x86
    Build JNIBridge x86

build.pl jnibridge x86_64
    Build JNIBridge x86_64

build.pl projectfiles
    Generate Visual Studio project files

build.pl test
    Test JNIBridge (only works on OSX)

build.pl help
    Show command line arguments

```

Note: Things left to do, I think it's wrong that Bee currently creates Zip pacakge, because Yamato also creates zip package, thus you have a zip with zip inside, which is lame, but didn't want to fix this in this PR

### CODE REVIEW

File: .yamato/build.yml [hunk 1]
```diff
@@ -1,14 +1,28 @@
...
+        - build/builds.zip
+
+build_for_android_win:
+  name: Build JNIBridge Windows (For Local Development)
+  agent:
+    type: Unity::VM
+    image: cds-ops/win10-base:stable
+    flavor: b1.xlarge
+  variables:
+    ANDROID_SDK_ROOT: C:\Users\tomas\android-sdk_auto
```
Comment (aurimasc) [inline .yamato/build.yml]: Maybe it shouldn't be your name :)

File: JniBridge.bee.cs [hunk 2]
```diff
@@ -394,4 +407,80 @@ static void SetupTestProgramOsx(ToolChain toolchain, NativeProgram staticLib, Co
...
+            return "x86_64";
+        throw new NotImplementedException(architecture.ToString());
+    }
+
+    static void SetupGeneratedProjects(IReadOnlyList<NativeProgram> programs)
+    {
+        var sln = new VisualStudioSolution();
+
+
+        // Not entirely sure, but I think there's should have been only NativeProgram with different architectures.
```
Comment (aurimasc) [inline JniBridge.bee.cs:504]: The reason would be my lack of knowledge about bee, it was the first time I seriously did something with it, so feel free to change it.

**Review by aurimasc — CHANGES_REQUESTED** [review]
Few things:
- If you want to develop on Windows, you for sure want to setup test program to run on Windows. That's the fastest way to test your changes.
- Why did you commit 7zip? I can't easily find it being used anywhere?
- Have you made sure that bee.bat & bee.ps1 are in sync with bee shell script (as in  the same bee version)
- I don't like putting stuff into perl. I would have removed it completely had required JDK/NDK were available in Stevedore. I really suggest making everything bee-centric so we're ready to kill perl with fire with least amount changes.

File: .yamato/build.yml (hunk 1 above)
Comment (todi1856) [reply .yamato/build.yml]: whoops

File: JniBridge.bee.cs (hunk 2 above)
Comment (todi1856) [reply JniBridge.bee.cs:504]: sadly, I don't know too, and I was afraid of changing your build code :)

File: .yamato/build.yml (hunk 1 above)
Comment (todi1856) [reply .yamato/build.yml]: > Few things:
> 
> * If you want to develop on Windows, you for sure want to setup test program to run on Windows. That's the fastest way to test your changes.
Will try to implement it.
> * Why did you commit 7zip? I can't easily find it being used anywhere?
Used here - https://github.com/Unity-Technologies/android-jni-bridge/blob/b789a5a8154ed66cfdb6e06d28c9f7932fcca89f/PrepareAndroidSDK.pm#L151
> * Have you made sure that bee.bat & bee.ps1 are in sync with bee shell script (as in  the same bee version)
Yes, I did, but I'll see if OSX/Windows can share a file
> * I don't like putting stuff into perl. I would have removed it completely had required JDK/NDK were available in Stevedore. I really suggest making everything bee-centric so we're ready to kill perl with fire with least amount changes
I didn't add any building code in perl, I only added the interactivity element, so you wouldn't need to know what to pass. Bee doesn't allow you to do that, so unlikely you will be able to remove perl completely

Comment (aurimasc) [reply .yamato/build.yml]: Making tests run on windows shouldn't be difficult, just need to replace darwin specific paths/libs with win ones.
In unity repo "jam --help AndroidPlayer" prints out the info quite nicely and I thing that info comes from bee, so I guess it shouldn't be too difficult to hook a help into bee. That won't be interactive, but printing out commands IMO is good enough.

**Comment by todi1856** [comment]
@aurimasc please check latest changes, sadly I couldn't make JNIBridge compile with Visual Studio, see this comment
https://github.com/Unity-Technologies/android-jni-bridge/pull/36/commits/366a16d05ce261b6624712b999c112416e9d9729

So for now I'll keep tests on Windows disabled, Visual Studio doesn't like this line:

```
template <typename T>
class Array : public ObjectArray<T>
{
public:
	explicit inline Array(jobject obj)                         : ObjectArray<T>(obj) {};
	explicit inline Array(jobjectArray obj)                    : ObjectArray<T>(obj) {};
	explicit inline Array(size_t length, T initialElement = 0) : ObjectArray<T>(T::__CLASS, length, initialElement) {}; <-------cannot resolve __CLASS
	template<typename T2>
	explicit inline Array(size_t length, T2* elements)         : ObjectArray<T>(T::__CLASS, length, elements) {};
};

```

There are many more issues like that

File: JniBridge.bee.cs [hunk 3]
```diff
@@ -37,6 +39,12 @@ class JniBridge
 {
     const string kAndroidApi = "android-31";
 
+    class Platform
```
Comment (aurimasc) [inline JniBridge.bee.cs]: I think this class should be static

File: APIHelper.h [hunk 4]
```diff
@@ -2,6 +2,33 @@
...
+#ifdef GetFreeSpace
+#undef GetFreeSpace
+#endif
+#ifdef Yield
+#undef Yield
+#endif
+
+inline int __sync_add_and_fetch(int volatile* i, int value)
+{
+	return _InterlockedExchangeAdd((long volatile*)i, value) + value;
```
Comment (aurimasc) [inline APIHelper.h:19]: Add a static assert somewhere that int and long are of the same size

File: JNIBridge.cpp [hunk 5]
```diff
@@ -1,26 +1,47 @@
...
+#ifdef WINDOWS
+#include <windows.h>
+#else
 #include <pthread.h>
+#endif
 
 namespace jni
 {
 
 // __thread is not supported on android (dynamic linker) :(
```
Comment (aurimasc) [inline JNIBridge.cpp:13]: It's possible that this comment is out of date. Worth to try C++11 "thread_local" keyword:
https://en.cppreference.com/w/c/thread/thread_local

File: Proxy.cpp [hunk 6]
```diff
@@ -105,29 +105,37 @@ void ProxyObject::DisableInstance(jobject proxy)
...
 	pthread_mutex_init(&lock, &attr);
 	pthread_mutexattr_destroy(&attr);
+#endif
 	head = NULL;
 }
 
 ProxyTracker::~ProxyTracker()
 {
+#ifdef WINDOWS
+	DeleteCriticalSection(&lock);
```
Comment (aurimasc) [inline Proxy.cpp]: Maybe try using mutex from standard library? This code probably predates existence of it.

**Comment by aurimasc** [comment]
> 	explicit inline Array(size_t length, T initialElement = 0) : ObjectArray<T>(T::__CLASS, length, initialElement) {}; <-------cannot resolve __CLASS
...
> There are many more issues like that

The __CLASS is in generated code. For work in VS you may need to run generator at least once.

**Comment by todi1856** [comment]
> > explicit inline Array(size_t length, T initialElement = 0) : ObjectArray(T::__CLASS, length, initialElement) {}; <-------cannot resolve __CLASS
> > ...
> > There are many more issues like that
> 
> The __CLASS is in generated code. For work in VS you may need to run generator at least once.

I did run it, the same issue occurs in non generated code:

2>E:\Projects\android-jni-bridge\Proxy.h(78): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax

```
template <class RefAllocator, class ...TX>
class ProxyGenerator : public ProxyObject, public TX::__Proxy...
{	
protected:
	ProxyGenerator() : m_ProxyObject(NewInstance(this, (jobject[]){TX::__CLASS...}, sizeof...(TX)))	
	{
		proxyTracker.StartTracking(this);
	}
```

**Comment by aurimasc** [comment]
> > > explicit inline Array(size_t length, T initialElement = 0) : ObjectArray(T::__CLASS, length, initialElement) {}; <-------cannot resolve __CLASS
> > > ...
> > > There are many more issues like that
> > 
> > 
> > The __CLASS is in generated code. For work in VS you may need to run generator at least once.
> 
> I did run it, the same issue occurs in non generated code:

Maybe try to rename it to something that does not have leading underscore? Leading underscores I kind of reserved for compiler use, may have problems when you are switching compilers.

> 2>E:\Projects\android-jni-bridge\Proxy.h(78): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax

This error sounds very much like compiler understands the code but refuses it due to it being not conformant with the standard. Maybe there is a compiler flag to enable it?

**Comment by todi1856** [comment]
> Maybe try to rename it to something that does not have leading underscore? Leading underscores I kind of reserved for compiler use, may have problems when you are switching compilers.

__CLASS is legal in Visual Studio. The problem is Visual Studio is not capable of compiling a template written like this
It's not about this, it seems like Visual Studio has much worse template parsing than clang

> 
> > 2>E:\Projects\android-jni-bridge\Proxy.h(78): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax
> 
> This error sounds very much like compiler understands the code but refuses it due to it being not conformant with the standard. Maybe there is a compiler flag to enable it?

I tried searching with no luck, I could rewrite code, but seeing that's there's huge template compile issue. Not sure if it's worth it.

I'll try to poke more a bit, but I almost ready to give up on tests on Windows

**Comment by aurimasc** [comment]
> __CLASS is legal in Visual Studio. The problem is Visual Studio is not capable of compiling a template written like this It's not about this, it seems like Visual Studio has much worse template parsing than clang

Maybe by change we have a Windows version of clang available in Stevedore?

> I'll try to poke more a bit, but I almost ready to give up on tests on Windows

Not being able to run tests means you'd have to build for android, then copy over to unity to build there and test the actual player with all the extra complexity that adds. Having to work on Mac can't be THAT bad :)
It is a lot better to test your codegen by building and running a test program, leaving android player for the last when you have completed the codegen part.

Unrelated: all these __CLASS look like they may have to be changed to support your scenario with GameActivity. Your problems mostly come from generated code, so if you find how to fix templates, there shouldn't be all that much changes.

**Comment by todi1856** [comment]
> > __CLASS is legal in Visual Studio. The problem is Visual Studio is not capable of compiling a template written like this It's not about this, it seems like Visual Studio has much worse template parsing than clang
> 
> Maybe by change we have a Windows version of clang available in Stevedore?

We don't. We only support building via Visual Studio on Windows...

> 
> > I'll try to poke more a bit, but I almost ready to give up on tests on Windows
> 
> Not being able to run tests means you'd have to build for android, then copy over to unity to build there and test the actual player with all the extra complexity that adds. Having to work on Mac can't be THAT bad :) It is a lot better to test your codegen by building and running a test program, leaving android player for the last when you have completed the codegen part.

Huh? The tests will still run on Yamato, only iterating tests on Windows is not possible

> 
> Unrelated: all these __CLASS look like they may have to be changed to support your scenario with GameActivity. Your problems mostly come from generated code, so if you find how to fix templates, there shouldn't be all that much changes.

I don't understand, how's that relevent. __CLASS is a simple variable name. For game activity, I simply need a different way to get jclass from string, this doesn't require any changes to __CLASS variable.

**Comment by aurimasc** [comment]
> > Unrelated: all these __CLASS look like they may have to be changed to support your scenario with GameActivity. Your problems mostly come from generated code, so if you find how to fix templates, there shouldn't be all that much changes.
> 
> I don't understand, how's that relevent. __CLASS is a simple variable name. For game activity, I simply need a different way to get jclass from string, this doesn't require any changes to __CLASS variable.

It's a static variable, so it might be prone to all your class-loader related issues.

File: APIHelper.h (hunk 4 above)
Comment (aurimasc) [reply APIHelper.h:19]: I just remembered one thing that is worth to try.
I've tried replacing the custom ref counter with std::shared_ptr with custom deleter, but that did not compile. Then, when adding bee support I've found API level burried in perl/make build code. That might be the reason why some modern C++ standard library feature were not available in JNIBridge, while work perfectly fine in unity.
If std::shared_ptr compiles, then using it with custom deleter would would cleanup the code and you won't need these ugly replacements for clang intrinsics.

File: JNIBridge.cpp [hunk 7]
```diff
@@ -1,26 +1,47 @@
 #include "JNIBridge.h"
 #include <stdlib.h>
 #include <string.h>
+#ifdef WINDOWS
+#include <windows.h>
```
Comment (aurimasc) [inline JNIBridge.cpp:5]: Here the same thing: try using mutex from C++ standard library.

File: .yamato/build.yml (hunk 1 above)
Comment (todi1856) [reply .yamato/build.yml]: bee --help is already taken by bee specific help, you cannot use that for your own usage.

File: JniBridge.bee.cs (hunk 3 above)
Comment (todi1856) [reply JniBridge.bee.cs]: fixed

File: .yamato/build.yml (hunk 1 above)
Comment (aurimasc) [reply .yamato/build.yml]: There is a Help attribute, I see it used in AndroidSupport.cs

File: APIHelper.h (hunk 4 above)
Comment (todi1856) [reply APIHelper.h:19]: std::shared_ptr compiles for JNIBridge android, but I would like to avoid adding extra changes on Android side, since it might be dangerous

File: JNIBridge.cpp (hunk 5 above)
Comment (todi1856) [reply JNIBridge.cpp:13]: Same as above, I would really like to avoid changing the functionality on Android side, just for the sake of having less code

File: Proxy.cpp (hunk 6 above)
Comment (todi1856) [reply Proxy.cpp]: Same as above, I would really like to avoid changing the functionality on Android side, just for the sake of having less code

File: APIHelper.h (hunk 4 above)
Comment (todi1856) [reply APIHelper.h:19]: added static assert

File: Proxy.cpp (hunk 6 above)
Comment (aurimasc) [reply Proxy.cpp]: This one I disagree with. You already changed locking code adding use of critical section instead of pthread_mutex on windows. std::mutex would remove this complexity. And it is a mutex.
IIRC, DOTS/Tiny Android code used std::mutex.

Comment (todi1856) [reply Proxy.cpp]: We don't ship windows, we ship Android, you're suggesting to change Android code which directly affects users. I am worried that if there's a difference with how std::mutex works when compared with pthread mutex, if there is at least some difference, that's already a red flag.

@juliusmik any comments on this one?

File: .yamato/build.yml (hunk 1 above)
Comment (todi1856) [reply .yamato/build.yml]: The Help attribute comes form Unity build system code, which only lives inside unity repo Tools\Unity.BuildSystem\Unity.BuildSystem\Help.cs

it's not a part of default Bee setup. I looked a bit, and it seems it's quite big amount of code, just to produce few lines of help - Tools\Unity.BuildSystem\Unity.BuildSystem\Main.cs. That's just not worth it for JNIBridge

File: Proxy.cpp (hunk 6 above)
Comment (aurimasc) [reply Proxy.cpp]: The use of mutex in ProxyTracker is the most trivial one. We even create a recursive mutex now for no reason. I think you can safely assume std::mutex does lock/unlock.

File: .yamato/build.yml (hunk 1 above)
Comment (todi1856) [reply .yamato/build.yml]: p.s there's bee -t command, but besides useful commands, it prints lots of garbage
```
E:\Projects\android-jni-bridge>bee -t

Named nodes and aliases:
----------------------------------------------------------------
 - all_tundra_nodes
 - apigenerator
 - build:android:arm64-v8a
 - build:android:armeabi-v7a
 - build:android:x86
 - build:android:x86_64
 - build:android:zip
 - build:osx:test
 - build:osx:x86_64
 - build:windows:test
 - build:windows:x86_64
 - generate:jnibridge:android
 - generate:jnibridge:osx
 - generate:jnibridge:windows
 - includes:android
 - jnibridge
 - objectFileDependencyMetaNode_JNIBridgeTests_debug_Mac_x64_nonlump
 - objectFileDependencyMetaNode_JNIBridgeTests_debug_Win64_VS2019_nonlump
 - objectFileDependencyMetaNode_android-31_release_Android_arm32_nonlump
 - objectFileDependencyMetaNode_android-31_release_Android_arm64_nonlump
 - objectFileDependencyMetaNode_android-31_release_Android_x64_nonlump
 - objectFileDependencyMetaNode_android-31_release_Android_x86_nonlump
 - objectFileDependencyMetaNode_testlib_debug_Mac_x64_nonlump
 - objectFileDependencyMetaNode_testlib_debug_Win64_VS2019_nonlump
 - projectfiles
```

File: JniBridge.bee.cs [hunk 8]
```diff
@@ -1,397 +1,610 @@
-using System;
```
Comment (juliusmik) [inline JniBridge.bee.cs]: what happened here? why is this file not diffed properly?

File: build.pl [hunk 9]
```diff
@@ -1,118 +1,325 @@
-#!/usr/bin/env perl -w
```
Comment (juliusmik) [inline build.pl]: Same here, why isn't it diffed properly?

File: test/Test.cpp [hunk 10]
```diff
@@ -20,11 +59,12 @@ int main(int,char**)
 
 	char classPath[] = {"-Djava.class.path=build/jnibridge.jar"};
 
-	JavaVMOption options[1];
+	JavaVMOption options[3];
```
Comment (juliusmik) [inline test/Test.cpp]: should be 2?
Maybe it would be worth to create a constant for this, especially since we specify "vm_args.nOptions = 2;" as well below.

File: test/Test.cpp [hunk 11]
```diff
@@ -120,41 +164,51 @@ int main(int,char**)
 	java::lang::Object object = java::lang::Integer(23754);
 	if (jni::InstanceOf<java::lang::Number>(object) && jni::Cast<java::lang::Number>(object))
 	{
-		printf("%d\n", static_cast<int>(java::lang::Number(object)));
+		printf("%d\n", static_cast<jint>(java::lang::Number(object)));
 	}
 	else
 	{
-		int* p = 0; *p = 3;
+		exit(-1);
```
Comment (juliusmik) [inline test/Test.cpp]: why exit wasn't used before? maybe exit is not caught as a failure on osx, so null pointer was used? I'm only guessing. This probably needs to be tested to make sure it is counted as a failure on OSX.

**Review by juliusmik — COMMENTED** [review]
* I really like the informative output when running build.pl.
* I agree with Aurimas, that committing 7z to this repo is questionable, since it's only used for windows. Maybe we could do similarly to Unity repo, where we have Tools/Bee/manifest.stevedore file and specify "public: 7za-win-x64/38c5b39be2e8_a333cfccb708c88459b3812eb2597ca486ec9b416172543ca3ef8e5cd5f80984.zip" in it to allow be to download it? We would probably need to run bee for that before calling PrepareAndroidSDK, but as Aurimas has said, we would want to get rid of that whole PrepareAndroidSDK file once the required SDK and JDK versions are available in artifactory.
* Regarding the sdk::mutex, I would expect it to use pthread_mutex on Android, so it should be safe to use sdk::mutex. You can consult with the platform-foundation team (if that's how they are still called), since they've worked on some unified synchronisation implementations in baselib.

File: JniBridge.bee.cs (hunk 8 above)
Comment (todi1856) [reply JniBridge.bee.cs]: end lines are not configured for this repo....will fix

File: test/Test.cpp (hunk 11 above)
Comment (todi1856) [reply test/Test.cpp]: with that *p = 3 hack, the failure would look like this
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010f6a0837, pid=2125, tid=0x0000000000000103
#
# JRE version: OpenJDK Runtime Environment (8.0) (build 1.8.0-adoptopenjdk-jenkins_2018_05_19_02_01-b00)
# Java VM: OpenJDK 64-Bit Server VM (25.71-b00 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [JNIBridgeTests+0x1837]  main+0x877
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/tomasdirvanauskas/Projects/android-jni-bridge/hs_err_pid2125.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Test failed at ./build.pl line 289, <CONFIG> line 2.
```

with exit (-1); 

you would get only this
`
Test failed at ./build.pl line 289, <CONFIG> line 2.`

which is much cleaner, but I see it's lacking info, i'll switch to assert instead, since provide line numbers

**Comment by todi1856** [comment]
@aurimasc @juliusmik could you give it another go.

Removed 7z, fixed up asserts to properly report message and errors.

It will fail with, example
'
Failures related to AttachCurrentThread at line 119
Exited with code -1
'

File: test/Test.cpp [hunk 12]
```diff
@@ -120,41 +176,50 @@ int main(int,char**)
 	java::lang::Object object = java::lang::Integer(23754);
 	if (jni::InstanceOf<java::lang::Number>(object) && jni::Cast<java::lang::Number>(object))
 	{
-		printf("%d\n", static_cast<int>(java::lang::Number(object)));
+		printf("%d\n", static_cast<jint>(java::lang::Number(object)));
```
Comment (aurimasc) [inline test/Test.cpp]: Why this change? %d specifier means int and it's not guaranteed that jint is typedefed to int.

File: test/Test.cpp [hunk 13]
```diff
@@ -120,41 +176,50 @@ int main(int,char**)
...
 	// Array Test
 	// -------------------------------------------------------------
 	{
 		jni::LocalScope frame;
-		jni::Array<int> test01(4, (int[]){1, 2, 3, 4});
+		jint ints[] = { 1, 2, 3, 4 };
+		jni::Array<jint> test01(4, ints);
 		for (int i = 0; i < test01.Length(); ++i)
-			printf("ArrayTest01[%d],", test01[i]);
+			printf("ArrayTest01[%d],", (jint)test01[i]);
```
Comment (aurimasc) [inline test/Test.cpp]: Same here: int vs jint

**Review by aurimasc — COMMENTED** [review]
int vs jint need to be gone through.

File: test/Test.cpp (hunk 12 above)
Comment (todi1856) [reply test/Test.cpp]: java::lang::Number doesn't have operator for casting to int
```
struct Number : ::java::lang::Object
{
	static jni::Class __CLASS;

	operator ::java::io::Serializable();
	::jbyte ByteValue() const;
	::jdouble DoubleValue() const;
	::jfloat FloatValue() const;
	::jint IntValue() const;
	::jlong LongValue() const;
	::jshort ShortValue() const;
	static jobject __Constructor();
	Number() : ::java::lang::Object(__Constructor()) { __Initialize(); }
	explicit Number(jobject o) : ::java::lang::Object(o) { __Initialize(); }
	Number(const Number& o)  : ::java::lang::Object(o) { __Initialize(); }
Number(Number&& o) = default;
Number& operator=(const Number& o) = default;
Number& operator=(Number&& o) = default;
operator ::jbyte() const;
operator ::jshort() const;
operator ::jint() const;
operator ::jlong() const;
operator ::jfloat() const;
operator ::jdouble() const;
private:
	void __Initialize();

};
```

thus you would get
1>E:\Projects\android-jni-bridge\test\Test.cpp(179): error C2440: 'static_cast': cannot convert from 'java::lang::Number' to 'int'

File: test/Test.cpp (hunk 13 above)
Comment (todi1856) [reply test/Test.cpp]: here it's indeed not needed anymore

File: test/Test.cpp (hunk 12 above)
Comment (aurimasc) [reply test/Test.cpp]: Sure, it needs double casting then. What I don't like is that the code assumes jint and int to be of the same size. jint is 32-bit, but int is not guaranteed to be 32-bit (although usually is).
Strange, that MS compiler treats the two as different types.

Comment (todi1856) [reply test/Test.cpp]: I think it's very good that MS complains about it, since jint even though under hood the same, is declared as new type.

So, can you explain, if you want something to fix here, or leave it as it is ?

Comment (aurimasc) [reply test/Test.cpp]: Checked our OpenJDK and on Mac jint is typedefed as int, while on Windows it's typedefed as long.
So you need to either double cast or cast to something universal (maybe jlong is universal?).

Comment (todi1856) [reply test/Test.cpp]: Fixed it, casting to jlong now

File: test/Test.cpp [hunk 14]
```diff
@@ -120,41 +176,50 @@ int main(int,char**)
 	java::lang::Object object = java::lang::Integer(23754);
 	if (jni::InstanceOf<java::lang::Number>(object) && jni::Cast<java::lang::Number>(object))
 	{
-		printf("%d\n", static_cast<int>(java::lang::Number(object)));
+		printf("%d\n", static_cast<jlong>(java::lang::Number(object)));
```
Comment (aurimasc) [inline test/Test.cpp]: It's not correct. You shouldn't use any of j* types, but rather C++ types. Try long, if that works on all platforms, then printf specifier should be %ld
https://www.cplusplus.com/reference/cstdio/printf/

Comment (todi1856) [reply test/Test.cpp]: Fixed again

Comment (aurimasc) [reply test/Test.cpp]: Unfortunately, I can find many other places where jint is outputed as %d.

Comment (todi1856) [reply test/Test.cpp]: check now

File: Proxy.h [hunk 15]
```diff
@@ -54,15 +56,21 @@ class ProxyTracker
...
 protected:
-	ProxyGenerator() : m_ProxyObject(NewInstance(this, (jobject[]){TX::__CLASS...}, sizeof...(TX)))	
+	ProxyGenerator() 
+#if WINDOWS
+	: m_ProxyObject(NewInstance(this, TX::__CLASS...))
 	{
+#else
+	: m_ProxyObject(NewInstance(this, (jobject[]){TX::__CLASS...}, sizeof...(TX)))	
+	{
+#endif
```
Comment (aurimasc) [inline Proxy.h]: Can you at least create a task to clean this up. I really don't like having separate code paths where can be only one.
In this case you can pass additional nullptr to NewInstance at the end and have it figure out the number of interfaces via va_list.

File: test/Test.cpp [hunk 16]
```diff
@@ -120,54 +176,65 @@ int main(int,char**)
...
-			printf("ArrayTest04[%d],", java::lang::Integer(test04[i]).IntValue());
+			printf("ArrayTest04[%ld],", (long)java::lang::Integer(test04[i]).IntValue());
 		printf("\n");
 
-		jni::Array<int> test05(4, (java::lang::Integer[]){1, 2, 3, 4});
+		java::lang::Integer integers4[] = { 1, 2, 3, 4 };
+		jni::Array<jint> test05(4, integers4);
 		for (int i = 0; i < test05.Length(); ++i)
-			printf("ArrayTest05[%d],", test05[i]);
+			printf("ArrayTest05[%ld],", (jint)test05[i]);
```
Comment (aurimasc) [inline test/Test.cpp:228]: Wrong here, should be long

File: Proxy.h (hunk 15 above)
Comment (todi1856) [reply Proxy.h]: Agreed, @aurimasc I can actually remove #else block and leave it how it's written for windows.

It seems to compile fine - https://github.com/Unity-Technologies/android-jni-bridge/pull/36/commits/d63fa4b827907520f1fef8e4e7cf3d848f90ffca

File: Proxy.h [hunk 17]
```diff
@@ -54,14 +56,14 @@ class ProxyTracker
...
-	pthread_mutex_t lock;
+	std::mutex lock;
 };
 
 template <class RefAllocator, class ...TX>
 class ProxyGenerator : public ProxyObject, public TX::__Proxy...
 {	
 protected:
-	ProxyGenerator() : m_ProxyObject(NewInstance(this, (jobject[]){TX::__CLASS...}, sizeof...(TX)))	
+	ProxyGenerator() : m_ProxyObject(NewInstance(this, TX::__CLASS...))
```
Comment (aurimasc) [inline Proxy.h]: This last change is not correct. By it you limit the number of interfaces implemented by proxy to 2.
You need to pass nullptr as additional argument after "TX::__CLASS..." and have:
static jobject NewInstance(void* nativePtr, ...);
where you would use va_list to go over extra args until you encounter nullptr.

**Review by aurimasc — CHANGES_REQUESTED** [review]

File: Proxy.h (hunk 17 above)
Comment (aurimasc) [reply Proxy.h]: Alternatively it can be a recursive variadic function template that fills an array. But I personally would go with va_list instead.

Comment (todi1856) [reply Proxy.h]: I think I have even better variant, no recursion and compiles on Windows and Mac

https://github.com/Unity-Technologies/android-jni-bridge/commit/5b1486174506d78c4605fd1c9bdd63f8e86d5907

slightly fixed

https://github.com/Unity-Technologies/android-jni-bridge/commit/67396c057ce80c42e75bd92fedc9053bcd7618eb

**Review by aurimasc — APPROVED** [review]

File: JniBridge.bee.cs [hunk 18]
```diff
@@ -388,10 +443,168 @@ static void SetupTestProgramOsx(ToolChain toolchain, NativeProgram staticLib, Co
...
+        var javaLibDir = jdk.JavaHome.Combine("lib");
+        np.Libraries.Add(new StaticLibrary(javaLibDir.Combine("jvm.lib")));
+
+        var destDir = new NPath("build").Combine(Platform.Windows);
+        var config = new NativeProgramConfiguration(codegen, toolchain, false);
+        var target = np.SetupSpecificConfiguration(config, config.ToolChain.ExecutableFormat).DeployTo(destDir);
+
+        var exeName = np.Name + ".exe";
+        var pdbName = np.Name + ".pdb";
+        var server = jdk.JavaHome.Combine("jre/bin/server");
```
Comment (juliusmik) [inline JniBridge.bee.cs]: Do I understand correctly that you copy files into the jdk directory (therefore polluting the jdk installation)? Why is this needed?
By the way, with JDK 11 the jre directory will not exist anymore.

Comment (aurimasc) [reply JniBridge.bee.cs]: I think it copies files in the other direction. Still, why?

File: JniBridge.bee.cs [hunk 19]
```diff
@@ -388,10 +443,168 @@ static void SetupTestProgramOsx(ToolChain toolchain, NativeProgram staticLib, Co
...
+        targetExecutable = Backend.Current.SetupCopyFile(server.Combine(exeName), destDir.Combine(exeName));
+        var targetPdb = Backend.Current.SetupCopyFile(server.Combine(pdbName), destDir.Combine(pdbName));
+        var targetJar = Backend.Current.SetupCopyFile(server.Combine("jnibridge.jar"), "build/jnibridge.jar");
```
Comment (aurimasc) [inline JniBridge.bee.cs]: Just noticed this: is this copying jnibridge.jar to JDK or the opposite why (meaning a non-existent file gets copyied)?

File: JniBridge.bee.cs (hunk 18 above)
Comment (todi1856) [reply JniBridge.bee.cs]: There were couple of reasons:
* Executable wouldn't run without jvm.dll, I tried copying it next the exe. But then I would get error that Java VM can't initialize
* The other reason it would start failing to find jnibridge.jar
* Finally I found that if you place and run exe from the server directory, it runs fine

But I think I found a slightly better solution, one sec...

Comment (aurimasc) [reply JniBridge.bee.cs]: Blind guesses:
- is JAVA_HOME set when you run tests?
- maybe you need class path entry for Java standard libs?

Comment (todi1856) [reply JniBridge.bee.cs]: Ok, here's my solution - https://github.com/Unity-Technologies/android-jni-bridge/pull/36/commits/a736cbf8d15db719c4f1af06a0f0dc9900fe6b6b

Don't know about JAVA_HOME, probably not set. But I probably wouldn't want to set it anyways, since it would conflict with my JAVA on PC.

Comment (aurimasc) [reply JniBridge.bee.cs]: So the problem was, that your program could not find jvm.dll, which makes sense. Copying it to other place expectedly did not work, because there's plenty of stuff in Java install and some of it is probably required to run.
But I think you want to set JAVA_HOME to what you actually use. I think that variable might be checked, so you risk problems when you have Java installed on machine and that variable set, but use a different one.

Comment (aurimasc) [reply JniBridge.bee.cs]: On a second thought, if you haven't changed things around, then bee script uses Java from JAVA_HOME and falls back to one from Stevedore, so it's probably fine then.

**Review by juliusmik — APPROVED** [review]
